File: 96-unique-binary-search-trees/96-unique-binary-search-trees.py
class Solution:
    def numTrees(self, n: int) -> int:
        numTree = [1] * (n+1)

        # 0 node = 1 tree
        # 1 node = 1 tree
        for nodes in range(2, n+1):
            total = 0
            for root in range(1, nodes+1):
                left = root-1
                right = nodes-root
                total += numTree[left]*numTree[right]
            numTree[nodes] = total
        return numTree[n]

File: 96-unique-binary-search-trees/96-unique-binary-search-trees.cpp
class Solution
{
public:
  int dp[20]{};
  int numTrees(int n)
  {
    // numTrees[4] =  numTrees[0] * numTrees[3] +
    //                numTrees[1] * numTrees[2] +
    //                numTrees[2] * numTrees[1] +
    //                numTrees[3] * numTress[1]
    if (n <= 1)
      return 1;
    if (dp[n])
      return dp[n];
    for (int i = 1; i <= n; i++)
    {
      dp[n] += numTrees(i - 1) * numTrees(n - i);
    }
    return dp[n];
  }
};